### PR TITLE
Implement backup item action plugin for volumesnapshots

### PR DESCRIPTION
**Pull request by ashish-amarnath**
Signed-off-by: Ashish Amarnath ashisham@vmware.com

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -13,8 +13,7 @@
 # limitations under the License.
```
Comment (ashish-amarnath) [inline Makefile]: Changes to this file are unrelated to this PR. They come from PR #29

File: internal/backup/csi_snapshotter.go [hunk 2]
```diff
@@ -20,17 +20,14 @@ import (
 	"fmt"
```
Comment (ashish-amarnath) [inline internal/backup/csi_snapshotter.go]: Changes to this file are unrelated to this PR. They come from #30

**Comment by ashish-amarnath** [comment]
Socializing the implementation of the backup item action plugin to backup volumesnapshotcontent as additional resource when backing up volumesnapshot.
Basic tests looked good. Going to take the changes in this PR through a little more testing.

**Comment by ashish-amarnath** [comment]
cc @xing-yang

**Comment by ashish-amarnath** [comment]
This change looked good on the regular backup and restore testing.
Need to run tests using the snapshot lister and snapshot delete secrets.

File: internal/backup/csi_volumesnapshot.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+// AppliesTo returns information indicating that the CSISnapshotter action should be invoked to backup volumesnapshots.
+func (p *VolumeSnapshotBackupItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	p.Log.Info("VolumeSnapshotBackupItemAction AppliesTo")
+
+	return velero.ResourceSelector{
+		IncludedResources: []string{"volumesnapshots.snapshot.storage.k8s.io"},
+	}, nil
+}
+
+func setVolumeSnapshotAnnotationsAndLabels(vs *snapshotv1beta1api.VolumeSnapshot, vals map[string]string) {
```
Comment (ashish-amarnath) [inline internal/backup/csi_volumesnapshot.go]: not a fan of duplicating the info in annotations and labels.

I would prefer to put all the metadata we generate into annotations, unless users may want to use the labels to query. WDYT?

**Comment by ashish-amarnath** [comment]
All testing looked good.

File: Makefile (hunk 1 above)
Comment (ashish-amarnath) [reply Makefile]: this PR got merged

File: internal/backup/csi_snapshotter.go (hunk 2 above)
Comment (ashish-amarnath) [reply internal/backup/csi_snapshotter.go]: This PR has been merged

File: internal/backup/csi_volumesnapshot.go (hunk 3 above)
Comment (skriss) [reply internal/backup/csi_volumesnapshot.go]: In general I've found labels to be more useful for expressing relationships, since you can query on them - for example, we use labels to relate PodVolumeBackups to Backups, which makes it easy to query all PodVolumeBackups that are related to a Backup for CLI display purposes, etc.

Other than that, if it's not something we're going to filter on, then I'd say an annotation make sense.

Comment (nrb) [reply internal/backup/csi_volumesnapshot.go]: I'd echo what Steve said - if it's something we might query/filter on, a label makes sense. Otherwise, it can be on an annotation.

I'm not sure if there's different storage limitations between the two or not.

File: internal/backup/csi_volumesnapshot.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+		util.CSIDriverNameLabel:        vsc.Spec.Driver,
+	}
+
+	additionalItems := []velero.ResourceIdentifier{}
+
+	// Capture the vsc's deletesnapshot secret annotations into the backup of the volumesnapshot
+	if util.IsVolumeSnapshotContentHasDeleteSecret(vsc) {
+		deleteSnapshotSecretName := vsc.Annotations[util.PrefixedSnapshotterSecretNameKey]
+		deleteSnapshotSecretNamespace := vsc.Annotations[util.PrefixedSnapshotterSecretNamespaceKey]
+		// TODO: add GroupResource for secret into kuberesource
```
Comment (nrb) [inline internal/backup/csi_volumesnapshot.go]: This was merged, though this repo's deps will have to be set to Velero's `master` if it's not already.

File: internal/backup/csi_snapshotter.go [hunk 5]
```diff
@@ -173,13 +174,24 @@ func (p *CSISnapshotter) Execute(item runtime.Unstructured, backup *velerov1api.
...
 			Name:          upd.Name,
 		},
-		{
-			GroupResource: kuberesource.VolumeSnapshotContents,
-			Name:          snapshotContent.Name,
-		},
 	}
 
-	p.Log.Debug("Listing additional items to backup")
+	if util.IsVolumeSnapshotClassHasListerSecret(snapshotClass) {
```
Comment (skriss) [inline internal/backup/csi_snapshotter.go]: Should this block of logic move into its own backup item action, that runs for VolumeSnapshotClasses? That way any time we back up a VolumeSnapshotClass, we back up its related secret. Also keeps this plugin simpler.

**Comment by carlisia** [comment]
> Cool. It's probably best to only add what we use. 👍

Sorry, I thought I was adding this comment to https://github.com/vmware-tanzu/velero/pull/2369. Please ignore.

File: internal/backup/csi_volumesnapshot.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+	vals := map[string]string{
+		util.VolumeSnapshotHandleLabel: *vsc.Status.SnapshotHandle,
+		util.CSIDriverNameLabel:        vsc.Spec.Driver,
+	}
+
+	additionalItems := []velero.ResourceIdentifier{}
+
+	// Capture the vsc's deletesnapshot secret annotations into the backup of the volumesnapshot
+	if util.IsVolumeSnapshotContentHasDeleteSecret(vsc) {
```
Comment (skriss) [inline internal/backup/csi_volumesnapshot.go]: Similar to the previous comment - Should this block of logic move into its own backup item action, that runs for VolumeSnapshotContent? That way any time we back up a VolumeSnapshotContent, we back up its related secret. Also keeps this plugin simpler.

File: internal/backup/csi_volumesnapshot.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+		vals[util.CSIDeleteSnapshotSecretName] = deleteSnapshotSecretName
+		vals[util.CSIDeleteSnapshotSecretNamespace] = deleteSnapshotSecretNamespace
+
+		p.Log.Infof("Found DeleteSnapshotSecret %s/%s on volumesnapshotcontent %s from its annotation", deleteSnapshotSecretNamespace, deleteSnapshotSecretName, vsc.Name)
+	} else {
+		p.Log.Infof("Volumesnapshotcontent %s does not have a deletesnapshot secret annotation", vsc.Name)
+	}
+
+	// save newly applied annotations and labels into the backed-up volumesnapshot
+	setVolumeSnapshotAnnotationsAndLabels(&vs, vals)
```
Comment (nrb) [inline internal/backup/csi_volumesnapshot.go]: This code copies the secret info off of the Class & Content objects and puts it on the VolumeSnapshot... when we restore, will this get copied from the snapshot back to the appropriate places?

The VolumeSnapshotClass should transfer over unchanged (at least, I think), but the VolumeSnapshotContents object will need to be recreated and the secret annotation added before the VolumeSnapshot is made, so either we need to read ahead and get the secret annotation off the VolumeSnapshot in the backup, or we need to store it another way, I think.

If https://github.com/vmware-tanzu/velero/pull/2323 is in, then we have all the VolumeSnapshots and VolumeSnapshotContents items stored in object storage as JSON arrays that could be downloaded without unpacking the whole tarball. That may help with order of operations.

File: internal/restore/volumesnapshot_restorer.go [hunk 8]
```diff
@@ -58,16 +62,61 @@ func (p *VSRestorer) Execute(input *velero.RestoreItemActionExecuteInput) (*vele
...
+		vsc.Annotations[util.PrefixedSnapshotterSecretNamespaceKey] = vs.Annotations[util.CSIDeleteSnapshotSecretNamespace]
 	}
 
-	if vsFromBackup.Status == nil || vsFromBackup.Status.BoundVolumeSnapshotContentName == nil {
-		return &velero.RestoreItemActionExecuteOutput{}, errors.Errorf("unable to lookup BoundVolumeSnapshotContentName from status")
+	_, snapClient, err := util.GetClients()
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	vscupd, err := snapClient.SnapshotV1beta1().VolumeSnapshotContents().Create(&vsc)
```
Comment (skriss) [inline internal/restore/volumesnapshot_restorer.go]: Did you consider returning the VSC as an additional item to restore, rather than directly creating it within the plugin? That's how we handle PVC->PV restores (see in-tree Velero plugins in pkg/restore). That's probably simpler, though I may be missing something.

File: internal/restore/volumesnapshot_restorer.go [hunk 9]
```diff
@@ -58,16 +62,61 @@ func (p *VSRestorer) Execute(input *velero.RestoreItemActionExecuteInput) (*vele
 		return &velero.RestoreItemActionExecuteOutput{}, errors.Wrapf(err, "failed to convert input.Item from unstructured")
 	}
 
-	var vsFromBackup snapshotv1beta1api.VolumeSnapshot
-	if err := runtime.DefaultUnstructuredConverter.FromUnstructured((input.ItemFromBackup.UnstructuredContent()), &vsFromBackup); err != nil {
-		return &velero.RestoreItemActionExecuteOutput{}, errors.Wrapf(err, "failed to convert input.ItemFromBackup from unstructured")
+	labels := vs.Labels
```
Comment (skriss) [inline internal/restore/volumesnapshot_restorer.go]: Can you add 1+ comments explaining the addition of L65-115? The method godoc should be updated, and any more detailed info can go inline.

File: internal/backup/csi_volumesnapshot.go (hunk 4 above)
Comment (ashish-amarnath) [reply internal/backup/csi_volumesnapshot.go]: I think waiting for a release to be better as it gives us more control over the changes with which this is built once this becomes more widely used. Updating the modules to use `master` may inadvertent build breaks or can result in unexpected behavior.

File: internal/backup/csi_volumesnapshot.go (hunk 6 above)
Comment (ashish-amarnath) [reply internal/backup/csi_volumesnapshot.go]: The current implementation doesn't explicitly add the VolumeSnapshotContent to the backup. But, now that I think about this, it might be better to have a restore item action plugin for VolumeSnapshotContents that does the conversion from dynamic snapshot to static snapshot.

File: internal/restore/volumesnapshot_restorer.go (hunk 8 above)
Comment (ashish-amarnath) [reply internal/restore/volumesnapshot_restorer.go]: The logic here is mainly converting the dynamic volumesnapshot to a static volumesnapshot.
Creating a new volumesnapshotcontent is what allows both originally/dynamically bound vsc and statically bound vsc to co-exist in the cluster.
Creating the vsc here also allows to setup the link between the volumesnapshot and volumesnapshotcontent. L118

Comment (nrb) [reply internal/restore/volumesnapshot_restorer.go]: I would have expected the creation of the new, static VSC to happen on restore. Velero's philosophy has usually been to back up things in the cluster more or less as-is, with minimal changes, then do any sort of manipulation/recreation on the restore side. That way, we have exactly what was in the cluster at backup time (even if it's something we created).

**Comment by nrb** [comment]
Here's how I envisioned the collection of plugins working - each plugin would be responsible for backing up it's own type from `AppliesTo`, possibly returning a type that would be processed by another plugin as an additional item.

To put it in concrete terms:

A `PersistentVolumeClaim` would generate a `VolumeSnapshot`, which would have it's own plugin.

A `VolumeSnapshot` would be backed up, and have a `VolumeSnapshotClass` and `VolumeSnapshotContents` as additional items.

The `VolumeSnapshotClass` would be backed up and have a Lister `Secret` as an additional item; I don't think there would be additional processing needed on the `VolumeSnapshotClass` here, just extracting the `Secret`, if any.

The `VolumeSnapshotContents` would be backed up and have the deletion `Secret` as an additional item.

This means each BackupItemAction plugin should contain fairly minimal logic, apart from the PVC one, which is creating a new `VolumeSnapshot`. The flip side is that the logic is then spread across each plugin instead of in one place, but I like that each plugin is responsible for it's own type. I'm not sure that extracting information from the `VSClass` and `VSContents` objects and appending it to the `VolumeSnapshot` is the right approach here.

However, on restore, we'll have more complicated logic, as we'll have to take the backed up `VolumeSnapshotContents` and extract information out and create a new one, rather than just feeding it in.

Additionally, I think Velero _should_ make a distinction between the CSI `VolumeSnapshots` it makes and those the user or other tools made. It _should_ treat them differently, because we don't want to mistakenly apply Velero logic to snapshots that some other tool is managing, or not apply Velero logic to snapshots that it should be managing. So in my opinion, we should focus this plugin work on CSI snapshots generated by Velero first, so that our users get a good integration experience.

**Comment by skriss** [comment]
I am very much in sync with @nrb's description of the various backup plugins and their functions in https://github.com/vmware-tanzu/velero-plugin-for-csi/pull/31#issuecomment-606862144, that seems like the logical breakout to me.

File: internal/restore/volumesnapshot_restorer.go [hunk 10]
```diff
@@ -58,16 +61,62 @@ func (p *VSRestorer) Execute(input *velero.RestoreItemActionExecuteInput) (*vele
...
+	// TODO: generated name will be like velero-velero-something. Fix that.
+	vsc := snapshotv1beta1api.VolumeSnapshotContent{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "velero-" + vs.Name + "-",
+			Annotations: map[string]string{
+				velerov1api.RestoreNameLabel: input.Restore.Name,
+			},
+		},
+		Spec: snapshotv1beta1api.VolumeSnapshotContentSpec{
+			DeletionPolicy: snapshotv1beta1api.VolumeSnapshotContentDelete,
```
Comment (ashish-amarnath) [inline internal/restore/volumesnapshot_restorer.go:83]: should this be `Delete` or `Retain`?

**Comment by skriss** [comment]
Changes this PR currently contains:
- add BackupItemAction for VolumeSnapshots that:
    - stores VolumeSnapshotContent attributes on the backed-up VolumeSnapshot, as annotations
    - returns VolumeSnapshotClass and VolumeSnapshotContents as additional items to back up
- add BackupItemAction for VolumeSnapshotClass that:
    - returns lister secret, if it exists, as additional item to back up
- modify RestoreItemAction for VolumeSnapshots to:
    - create new VolumeSnapshotContent and bind it to restored VolumeSnapshot
    - return delete secret, if it exists, as additional item to restore
- add RestoreItemAction for VolumeSnapshotClass that:
    -returns lister secret, if it exists, as additional item to restore

File: internal/restore/volumesnapshotclass_restorer.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
+/*
```
Comment (ashish-amarnath) [inline internal/restore/volumesnapshotclass_restorer.go:1]: This is pretty much a noop plugin atm.
Keeping this in this file as a place holder for now.

File: internal/backup/csi_volumesnapshotclass.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+func (p *VolumeSnapshotClassBackupItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	p.Log.Info("VolumeSnapshotClassBackupItemAction AppliesTo")
+
+	return velero.ResourceSelector{
+		IncludedResources: []string{"volumesnapshotclass.snapshot.storage.k8s.io"},
+	}, nil
+}
+
+// Execute backsup a VolumeSnapshotClass object and returns as additional items any snapshot lister secret that may be referenced in its annotations.
+func (p *VolumeSnapshotClassBackupItemAction) Execute(item runtime.Unstructured, backup *velerov1api.Backup) (runtime.Unstructured, []velero.ResourceIdentifier, error) {
```
Comment (ashish-amarnath) [inline internal/backup/csi_volumesnapshotclass.go:46]: This is a noop plugin atm.
Keeping this in this file as a place holder for the secret capturing PR that will follow this PR.

**Comment by ashish-amarnath** [comment]
This LGTM in testing.

**Comment by skriss** [comment]
👀

File: internal/backup/csi_volumesnapshot.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
+/*
```
Comment (skriss) [inline internal/backup/csi_volumesnapshot.go:1]: I'd like to clean up the file names to be a little clearer/consistent - so perhaps a standard of `<type>_action.go`, and this one becomes `volumesnapshot_action.go`?

File: internal/backup/csi_volumesnapshot.go [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+	if vs.Annotations == nil {
+		vs.Annotations = make(map[string]string)
+	}
+
+	for k, v := range vals {
+		vs.Annotations[k] = v
+	}
+}
+
+// Execute backsup a CSI volumesnapshot object and captures, as labels and annotations, information from its associated volumesnapshotcontents such as CSI driver name, storage snapshot handle
```
Comment (skriss) [inline internal/backup/csi_volumesnapshot.go:58]: s/backsup/backs up

File: internal/backup/csi_volumesnapshot.go [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+// AppliesTo returns information indicating that the VolumeSnapshotBackupItemAction should be invoked to backup volumesnapshots.
+func (p *VolumeSnapshotBackupItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	p.Log.Info("VolumeSnapshotBackupItemAction AppliesTo")
+
+	return velero.ResourceSelector{
+		IncludedResources: []string{"volumesnapshots.snapshot.storage.k8s.io"},
+	}, nil
+}
+
+func setVolumeSnapshotAnnotations(vs *snapshotv1beta1api.VolumeSnapshot, vals map[string]string) {
```
Comment (skriss) [inline internal/backup/csi_volumesnapshot.go:48]: there's nothing about this function that's specific to the VolumeSnapshot type, so you could change the signature to take a `metav1.ObjectMeta` and then be able to set the annotations of *any* Kubernetes type. Guessing there are spots for reuse across the other plugins too, though I haven't looked (yet).

File: internal/backup/csi_volumesnapshotclass.go [hunk 16]
```diff
@@ -0,0 +1,62 @@
+/*
```
Comment (skriss) [inline internal/backup/csi_volumesnapshotclass.go:1]: similar to prev comment, I'd call this `volumesnapshotclass_action.go`

File: internal/restore/volumesnapshot_restorer.go [hunk 17]
```diff
@@ -58,25 +61,68 @@ func (p *VSRestorer) Execute(input *velero.RestoreItemActionExecuteInput) (*vele
 		return &velero.RestoreItemActionExecuteOutput{}, errors.Wrapf(err, "failed to convert input.Item from unstructured")
 	}
 
-	var vsFromBackup snapshotv1beta1api.VolumeSnapshot
-	if err := runtime.DefaultUnstructuredConverter.FromUnstructured((input.ItemFromBackup.UnstructuredContent()), &vsFromBackup); err != nil {
-		return &velero.RestoreItemActionExecuteOutput{}, errors.Wrapf(err, "failed to convert input.ItemFromBackup from unstructured")
+	snapHandle, exists := vs.Annotations[util.VolumeSnapshotHandleLabel]
+	if !exists {
+		return nil, errors.Errorf("Volumesnapshot %s/%s does not have a %s label", vs.Namespace, vs.Name, util.VolumeSnapshotHandleLabel)
```
Comment (skriss) [inline internal/restore/volumesnapshot_restorer.go:66]: since it's being stored as an annotation, let's update the constant name and the log to say "annotation", not "label"

File: internal/util/util.go [hunk 18]
```diff
@@ -191,3 +191,45 @@ func GetClients() (*kubernetes.Clientset, *snapshotterClientSet.Clientset, error
...
 }
+
+// IsVolumeSnapshotClassHasListerSecret returns whether a volumesnapshotclass has a snapshotlister secret
+func IsVolumeSnapshotClassHasListerSecret(vc *snapshotv1beta1api.VolumeSnapshotClass) bool {
+	nameExists := false
+	nsExists := false
+
+	// https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/utils/util.go#L59-L60
+	// There is no release w/ these constants exported. Using the strings for now.
+	if vc.Annotations != nil {
```
Comment (skriss) [inline internal/util/util.go:202]: no need to nil-check this, since reads from nil maps are safe.

File: internal/util/util.go [hunk 19]
```diff
@@ -191,3 +191,45 @@ func GetClients() (*kubernetes.Clientset, *snapshotterClientSet.Clientset, error
 
 	return client, snapshotterClient, nil
 }
+
+// IsVolumeSnapshotClassHasListerSecret returns whether a volumesnapshotclass has a snapshotlister secret
+func IsVolumeSnapshotClassHasListerSecret(vc *snapshotv1beta1api.VolumeSnapshotClass) bool {
+	nameExists := false
```
Comment (skriss) [inline internal/util/util.go:197]: no need to declare these ahead of time, I'd just declare them as part of trying to read from the Annotations map.

File: internal/util/util.go [hunk 20]
```diff
@@ -191,3 +191,45 @@ func GetClients() (*kubernetes.Clientset, *snapshotterClientSet.Clientset, error
...
+		_, nameExists = vc.Annotations[PrefixedSnapshotterListSecretNameKey]
+		_, nsExists = vc.Annotations[PrefixedSnapshotterListSecretNamespaceKey]
+
+	}
+	return nameExists && nsExists
+}
+
+// IsVolumeSnapshotContentHasDeleteSecret returns whether a volumesnapshotcontent has a deletesnapshot secret
+func IsVolumeSnapshotContentHasDeleteSecret(vsc *snapshotv1beta1api.VolumeSnapshotContent) bool {
+	nameExists := false
```
Comment (skriss) [inline internal/util/util.go:212]: same 2 comments as for previous function

File: internal/util/util.go [hunk 21]
```diff
@@ -191,3 +191,45 @@ func GetClients() (*kubernetes.Clientset, *snapshotterClientSet.Clientset, error
...
+		_, nameExists = vsc.Annotations[PrefixedSnapshotterSecretNameKey]
+		_, nsExists = vsc.Annotations[PrefixedSnapshotterSecretNamespaceKey]
+	}
+	return nameExists && nsExists
+}
+
+// IsVolumeSnapshotHasVSCDeleteSecret returns whether a volumesnapshot should set the deletesnapshot secret
+// for the static volumesnapshotcontent that is created on restore
+func IsVolumeSnapshotHasVSCDeleteSecret(vs *snapshotv1beta1api.VolumeSnapshot) bool {
+	nameExists := false
```
Comment (skriss) [inline internal/util/util.go:227]: same 2 comments as for previous functions

File: internal/backup/csi_volumesnapshot.go (hunk 13 above)
Comment (ashish-amarnath) [reply internal/backup/csi_volumesnapshot.go:1]: Yes. That's on my list for cleanup along with the names that the plugins are registered with.
Will put out that cleanup PR after the once in flight have landed.

Comment (skriss) [reply internal/backup/csi_volumesnapshot.go:1]: OK - can you just file an issue for visibility?

File: internal/backup/csi_volumesnapshot.go (hunk 15 above)
Comment (ashish-amarnath) [reply internal/backup/csi_volumesnapshot.go:48]: Yes, this is also on my list for a cleanup. There are definitely spots for reuse, I can think of at least one.

Comment (ashish-amarnath) [reply internal/backup/csi_volumesnapshot.go:48]: Moved this function in this PR into the util package.

File: internal/backup/csi_volumesnapshot.go (hunk 13 above)
Comment (ashish-amarnath) [reply internal/backup/csi_volumesnapshot.go:1]: this file is being newly added in this PR. so it makes sense to rename this now.

File: internal/restore/volumesnapshot_restorer.go (hunk 10 above)
Comment (ashish-amarnath) [reply internal/restore/volumesnapshot_restorer.go:83]: Using delete for now. Will change later if required.

File: internal/backup/csi_volumesnapshot.go (hunk 4 above)
Comment (ashish-amarnath) [reply internal/backup/csi_volumesnapshot.go]: This will be in the imminent 1.3.2 release. Will fix up when updating the dependency

**Comment by ashish-amarnath** [comment]
@nrb this is ready for your 👀

File: internal/backup/volumesnapshot_action.go [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+// AppliesTo returns information indicating that the VolumeSnapshotBackupItemAction should be invoked to backup volumesnapshots.
+func (p *VolumeSnapshotBackupItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	p.Log.Info("VolumeSnapshotBackupItemAction AppliesTo")
+
+	return velero.ResourceSelector{
+		IncludedResources: []string{"volumesnapshots.snapshot.storage.k8s.io"},
+	}, nil
+}
+
+// Execute backsup a CSI volumesnapshot object and captures, as labels and annotations, information from its associated volumesnapshotcontents such as CSI driver name, storage snapshot handle
```
Comment (nrb) [inline internal/backup/volumesnapshot_action.go:48]: `backsup` -> `backs up`

Comment (ashish-amarnath) [reply internal/backup/volumesnapshot_action.go:48]: I thought I fixed that!

Comment (ashish-amarnath) [reply internal/backup/volumesnapshot_action.go]: fixed for real now!

File: internal/backup/volumesnapshot_action.go [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+	vsc, err := util.GetVolumeSnapshotContentForVolumeSnapshot(&vs, snapshotClient.SnapshotV1beta1(), p.Log)
+	if err != nil {
+		return nil, nil, errors.WithStack(err)
+	}
+
+	// Capture storage provider snapshot handle and CSI driver name
+	// to be used on restore to create a static volumesnapshotcontent that will be the source of the volumesnapshot.
+	vals := map[string]string{
+		util.VolumeSnapshotHandleAnnotation: *vsc.Status.SnapshotHandle,
+		util.CSIDriverNameAnnotation:        vsc.Spec.Driver,
```
Comment (nrb) [inline internal/backup/volumesnapshot_action.go:73]: This will need to be added to https://velero.io/docs/v1.3.2/csi/ - the driver _name_ will need to be the same on the destination cluster, not just the image/code.

File: internal/restore/volumesnapshot_restorer.go [hunk 24]
```diff
@@ -58,25 +61,68 @@ func (p *VSRestorer) Execute(input *velero.RestoreItemActionExecuteInput) (*vele
...
+		},
+	}
+
+	additionalItems := []velero.ResourceIdentifier{}
+
+	_, snapClient, err := util.GetClients()
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	// we create the volumesnapshotcontent here instead of relying on the restore flow because we want to statically
```
Comment (nrb) [inline internal/restore/volumesnapshot_restorer.go:102]: Thanks for this comment!

File: internal/util/util.go [hunk 25]
```diff
@@ -191,3 +191,40 @@ func GetClients() (*kubernetes.Clientset, *snapshotterClientSet.Clientset, error
 
 	return client, snapshotterClient, nil
 }
+
+// IsVolumeSnapshotClassHasListerSecret returns whether a volumesnapshotclass has a snapshotlister secret
```
Comment (nrb) [inline internal/util/util.go:195]: Minor nit: Period at the end for valid godoc string.

File: internal/util/util.go [hunk 26]
```diff
@@ -191,3 +191,40 @@ func GetClients() (*kubernetes.Clientset, *snapshotterClientSet.Clientset, error
...
+func IsVolumeSnapshotClassHasListerSecret(vc *snapshotv1beta1api.VolumeSnapshotClass) bool {
+	// https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/utils/util.go#L59-L60
+	// There is no release w/ these constants exported. Using the strings for now.
+	_, nameExists := vc.Annotations[PrefixedSnapshotterListSecretNameKey]
+	_, nsExists := vc.Annotations[PrefixedSnapshotterListSecretNamespaceKey]
+	return nameExists && nsExists
+}
+
+// IsVolumeSnapshotContentHasDeleteSecret returns whether a volumesnapshotcontent has a deletesnapshot secret
```
Comment (nrb) [inline internal/util/util.go:204]: Minor nit: Period at the end for valid godoc string.

File: internal/util/util.go [hunk 27]
```diff
@@ -191,3 +191,40 @@ func GetClients() (*kubernetes.Clientset, *snapshotterClientSet.Clientset, error
...
+func IsVolumeSnapshotContentHasDeleteSecret(vsc *snapshotv1beta1api.VolumeSnapshotContent) bool {
+	// https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/utils/util.go#L56-L57
+	// use exported constants in the next release
+	_, nameExists := vsc.Annotations[PrefixedSnapshotterSecretNameKey]
+	_, nsExists := vsc.Annotations[PrefixedSnapshotterSecretNamespaceKey]
+	return nameExists && nsExists
+}
+
+// IsVolumeSnapshotHasVSCDeleteSecret returns whether a volumesnapshot should set the deletesnapshot secret
```
Comment (nrb) [inline internal/util/util.go:213]: Minor nit: Period at the end for valid godoc string.

File: internal/util/util.go [hunk 28]
```diff
@@ -191,3 +191,40 @@ func GetClients() (*kubernetes.Clientset, *snapshotterClientSet.Clientset, error
...
+// for the static volumesnapshotcontent that is created on restore
+func IsVolumeSnapshotHasVSCDeleteSecret(vs *snapshotv1beta1api.VolumeSnapshot) bool {
+	_, nameExists := vs.Annotations[CSIDeleteSnapshotSecretName]
+	_, nsExists := vs.Annotations[CSIDeleteSnapshotSecretNamespace]
+	return nameExists && nsExists
+}
+
+// AddAnnotations adds the supplied key-values to the annotations on the object
```
Comment (nrb) [inline internal/util/util.go:221]: Minor nit: Period at the end for valid godoc string.

File: internal/util/util.go [hunk 29]
```diff
@@ -134,14 +134,14 @@ func GetVolumeSnapshotClassForStorageClass(provisioner string, snapshotClient sn
 	return nil, errors.Errorf("failed to get volumesnapshotclass for provisioner %s", provisioner)
 }
 
+// GetVolumeSnapshotContentForVolumeSnapshot returns the volumesnapshotcontent object associated with the volumesnapshot
```
Comment (nrb) [inline internal/util/util.go:137]: Minor nit: Period at the end for valid godoc string.

File: internal/restore/volumesnapshotclass_restorer.go [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+
+// AppliesTo returns information indicating that VolumeSnapshotClassRestoreItemAction should be invoked while restoring
+// volumesnapshotclass.snapshot.storage.k8s.io resrouces.
+func (p *VolumeSnapshotClassRestoreItemAction) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"volumesnapshotclass.snapshot.storage.k8s.io"},
+	}, nil
+}
+
+// Execute restores volumesnapshotclass objects returning any snapshotlister secret as additional items to restore
```
Comment (nrb) [inline internal/restore/volumesnapshotclass_restorer.go:42]: Minor nit: Period at the end for valid godoc string.

**Comment by ashish-amarnath** [comment]
I will add the `.` to update the go doc as part of a clean-up PR that is in the pipeline

File: internal/util/util.go (hunk 25 above)
Comment (ashish-amarnath) [reply internal/util/util.go:195]: added the `.` in https://github.com/vmware-tanzu/velero-plugin-for-csi/pull/37/commits/e4018c28326f338f6c346f6e8d40502d8c66ae7b

File: internal/util/util.go (hunk 26 above)
Comment (ashish-amarnath) [reply internal/util/util.go:204]: added the `.` in https://github.com/vmware-tanzu/velero-plugin-for-csi/pull/37/commits/e4018c28326f338f6c346f6e8d40502d8c66ae7b

File: internal/util/util.go (hunk 27 above)
Comment (ashish-amarnath) [reply internal/util/util.go:213]: added the `.` in https://github.com/vmware-tanzu/velero-plugin-for-csi/pull/37/commits/e4018c28326f338f6c346f6e8d40502d8c66ae7b

File: internal/util/util.go (hunk 28 above)
Comment (ashish-amarnath) [reply internal/util/util.go:221]: added the `.` in https://github.com/vmware-tanzu/velero-plugin-for-csi/pull/37/commits/e4018c28326f338f6c346f6e8d40502d8c66ae7b

File: internal/util/util.go (hunk 29 above)
Comment (ashish-amarnath) [reply internal/util/util.go:137]: added the `.` in https://github.com/vmware-tanzu/velero-plugin-for-csi/pull/37/commits/e4018c28326f338f6c346f6e8d40502d8c66ae7b

File: internal/restore/volumesnapshotclass_restorer.go (hunk 30 above)
Comment (ashish-amarnath) [reply internal/restore/volumesnapshotclass_restorer.go:42]: added the `.` in https://github.com/vmware-tanzu/velero-plugin-for-csi/pull/37/commits/e4018c28326f338f6c346f6e8d40502d8c66ae7b
